lib/Eisbach_Methods.thy

```diff
@@ -574,4 +575,156 @@ method if_concl for pattern::bool methods m =
+  assumes goal_multiplicity2: "\<And>n. \<lbrakk> P (n+10);P (n+20) \<rbrakk> \<Longrightarrow> P n"
+  assumes goal_multiplicity5: "\<And>n. \<lbrakk> P (n+1);P (n+2);P (n+3);P (n+4);P (n+5) \<rbrakk> \<Longrightarrow> P n"
+  assumes goal_multiplicity2x: "\<And>n. \<lbrakk> P n;Q n \<rbrakk> \<Longrightarrow> P n"
+  shows "P 0"
+proof -
+  (* tests for generating a three-deep stack of rule production *)
+
+  (* generating using semicolon *)
+  have "P 0"
+  apply (rule goal_multiplicity2; (print_concl, rule goal_multiplicity5
+         (* invoked only on P goals, results in 20 subgoals *)
+         ; (print_concl, rule goal_multiplicity2x)
+         (* does not change number of goals *)
+         ; (rule c)?))
+    by (repeat 20 \<open>rule a b\<close>)
+
+  have "P 0"
+    apply (rule goal_multiplicity2; rule c)
+    apply (rule goal_multiplicity5; rule c)
+    apply (fails \<open>rule goal_multiplicity2x; rule c\<close>)
+    apply (succeeds \<open>rule goal_multiplicity2x; (rule c)?\<close>)
+    by (repeat 6 \<open>rule a b\<close>)
+
+  (* generating using fwd_all_new - using ALLGOALS_REV and goal restriction *)
+  have "P 0"
+    apply (
+      fwd_all_new \<open>rule goal_multiplicity2\<close>
+        \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity5\<close>
+                                  \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>(rule c)?\<close>\<close>\<close>)
+    by (repeat 20 \<open>rule a b\<close>)
+
+  have "P 0"
+    apply (fwd_all_new \<open>rule goal_multiplicity2\<close> \<open>rule c\<close>)
+    apply (fwd_all_new \<open>rule goal_multiplicity5\<close> \<open>rule c\<close>)
+    apply (fails \<open>fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>rule c\<close>\<close>)
+    apply (succeeds \<open>fwd_all_new \<open>rule goal_multiplicity2x\<close> \<open>(rule c)?\<close>\<close>)
+    by (repeat 6 \<open>rule a b\<close>)
+
+  (* check also that we can solve everything it spits out *)
+  have "P 0"
+    by (
+      fwd_all_new
+        \<open>fwd_all_new \<open>rule goal_multiplicity2\<close>
+          \<open>print_concl, fwd_all_new \<open>rule goal_multiplicity5\<close>
+          \<open>rule goal_multiplicity2x\<close>\<close>\<close>
+        \<open>rule a b\<close>)
+  oops
+
+lemma "x = x"
+  apply (simp; fail)
+  done
+
+lemma "x = x"
+  apply (fwd_all_new simp fail)
+  done
+
+lemma "True"
+  apply (succeeds \<open>succeed; succeed\<close>)
+  apply (fails \<open>succeed; fail\<close>)
+  apply (fails \<open>fail; succeed\<close>)
+  apply (fails \<open>fail; fail\<close>)
+  apply (rule TrueI)
+  done
+
+lemma "True"
+  apply (succeeds \<open>fwd_all_new succeed succeed\<close>)
+  apply (fails \<open>fwd_all_new succeed fail\<close>)
+  apply (fails \<open>fwd_all_new fail succeed\<close>)
+  apply (fails \<open>fwd_all_new fail fail\<close>)
+  apply (rule TrueI)
+  done
+
+end (* fwd_all_new and fwd_all experiment *)
```

Comment:
nitpick: it's currently only really a `fwd_all_new` experiment

Reply:
True, we should add test cases for `fwd_all` as well.

Reply:
Added one experiment, not sure what else should go in there.